src/main/scala/workflow/graph/GraphExecutor.scala

```diff
@@ -0,0 +1,110 @@
+package workflow.graph
+
+/**
+ * A GraphExecutor is constructed using a graph with already-executed state attached to it.
+ * It provides methods to further execute parts of the graph, returning the execution result.
+ * By default, it will optimize the graph before any new execution occurs.
+ *
+ * Warning: Not thread-safe!
+ *
+ * @param graph The underlying graph to execute
+ * @param state A mapping of GraphId to computed Expression.
+ *              This is the current state of what has already been executed in the graph.
+ * @param optimize Whether or not to optimize the graph using the global executor before executing anything.
+ *                 Defaults to True.
+ */
+private[graph] class GraphExecutor(graph: Graph, state: Map[GraphId, Expression], optimize: Boolean = true) {
```

Comment:
Names are really confusing - what's the difference between a `GraphExecution` and a `GraphExecutor`?